src/main/scala-3/com/github/tminglei/slickpg/PgCompositeSupport.scala

```diff
@@ -0,0 +1,234 @@
+package com.github.tminglei.slickpg
+
+import izumi.reflect.macrortti.LightTypeTag
+import izumi.reflect.{Tag => TTag}
+
+import scala.reflect.{ClassTag, classTag}
+import composite.Struct
+import slick.jdbc.{PositionedResult, PostgresProfile}
+import slick.jdbc.SetParameter
+
+import scala.annotation.unchecked.uncheckedVariance
+import scala.deriving.*
+import scala.compiletime.{error, erasedValue, summonInline, summonFrom}
+import utils.PgTokenHelper._
+import utils.TypeConverters._
+import utils.RegisteredTypeConverter
+
+sealed abstract class BaseLevel[D <: BaseLevel[?, ?] , I <: BaseLevel[?, ?]] {
+  def i: Int
+  final type Dec = D
+  final type Inc = I
+  def -- : Dec
+  def ++ : Inc
+}
+sealed abstract class Level[D <: BaseLevel[?, ?] , I <: BaseLevel[?, ?]](val -- : D, val ++ : I, val i: Int) extends BaseLevel[D, I]
+case object Level_1 extends BaseLevel[Level_1.type, Level0.type]{
+  def i: Int = -1
+  override def -- = this
+  override def ++ = Level0
+}
+case object Level0 extends Level[Level_1.type, Level1.type](Level_1, Level1, 0)
+case object Level1 extends Level[Level0.type, Level2.type](Level0, Level2, 1)
+case object Level2 extends Level[Level1.type, Level3.type](Level1, Level3, 2)
+case object Level3 extends Level[Level2.type, Level4.type](Level2, Level4, 3)
+case object Level4 extends Level[Level3.type, Level5.type](Level3, Level5, 4)
+case object Level5 extends Level[Level4.type, Level6.type](Level4, Level6, 5)
+case object Level6 extends BaseLevel[Level5.type, Level6.type]{
+  def i: Int = 6
+  override def -- = Level5
+  override def ++ = this
+}
+
+
+case class Args(args: Any*)
+
+sealed trait TokenConverter[T, L <: BaseLevel[?, ?]](val l: L) {
+  type Type
+  type Level = L
+  type LevelInc = l.Inc
+  def levelInc: LevelInc = l.++
+  type LevelDec = l.Dec
+  def levelDec: LevelDec = l.--
+  def fromToken(token: Token): T
+
+  def toToken(value: T): Token
+}
+
+
+trait PgCompositeSupport extends utils.PgCommonJdbcTypes with array.PgArrayJdbcTypes { driver: PostgresProfile =>
```

Comment:
I have a feeling this file can be cleaned up a lot, but I ran out of steam, and this works for up to 6 levels of nesting which should cover most cases